src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java

```diff
@@ -25,28 +25,28 @@
  */
 package jdk.internal.classfile.impl;
 
-import java.lang.constant.ClassDesc;
-import java.lang.constant.MethodTypeDesc;
 import java.nio.ByteBuffer;
 import java.util.BitSet;
-import java.util.LinkedHashMap;
 import java.util.List;
-import java.util.stream.Collectors;
 import java.lang.classfile.TypeKind;
 import java.lang.classfile.constantpool.ConstantDynamicEntry;
 import java.lang.classfile.constantpool.DynamicConstantPoolEntry;
 import java.lang.classfile.constantpool.MemberRefEntry;
 import static java.lang.classfile.ClassFile.*;
+import java.lang.classfile.constantpool.Utf8Entry;
+import java.lang.constant.MethodTypeDesc;
+import java.util.ArrayDeque;
 
 
 public final class StackCounter {
 
+    private record Target(int bci, int stack) {}
+
     static StackCounter of(DirectCodeBuilder dcb, BufWriterImpl buf) {
         return new StackCounter(
                 dcb,
-                buf.thisClass().asSymbol(),
-                dcb.methodInfo.methodName().stringValue(),
-                dcb.methodInfo.methodTypeSymbol(),
+                dcb.methodInfo.methodName(),
+                dcb.methodInfo.methodType(),
```

Comment:
Can you enlighten me on why this actually leads to a performance improvement? Don't we already generate the methods with MethodTypeDesc symbols in ProxyGenerator so that they should be cached?

Reply:
This code is part of the **ClassFile API**’s internals, and so it doesn’t have access to `ProxyGenerator`’s cached `MethodTypeDesc`s, only the underlying `Utf8Entry`, so it’d need to be parsed.

Reply:
I see that now Class return type and Class array parameter types are directly converted to Strings instead of MTDs. I think that we should really run ProxyGenerator with JFR to profile and see the results (for instance, old ASM has a stackmap generation penalty due to parsing method descriptors on the fly compared to CF API, despite CF's slower overall performance due to allocations like Optional)

Reply:
These changes are based on profiling of the benchmarks.
StackCounter was not used in any performance critical applications and profiling revealed some obvious bottlenecks.